core/src/main/java/org/apache/calcite/rex/RexCall.java

```diff
@@ -204,7 +204,7 @@ public RelDataType getType() {
     // This reduction allows us to convert it to a semi-join.
     switch (getKind()) {
     case IS_NOT_NULL:
-      return !operands.get(0).getType().isNullable();
+      return RexSimplify.isSafeExpression(this) && !operands.get(0).getType().isNullable();
```

Comment:
we seem to be covered for this already by that the operators result is nullable even if all the inputs are non-nullable
```java
  @Test public void testUnsafeSimplify1() {
    assertFalse(vIntNotNull(0).getType().isNullable());
    assertTrue(plus(vIntNotNull(0), vIntNotNull(1)).getType().isNullable());
    RexCall c= (RexCall)isNotNull(plus(vIntNotNull(0), vIntNotNull(1)));
    assertFalse(c.isAlwaysFalse());
    assertFalse(c.isAlwaysTrue());
  }
```

Reply:
@kgyrtkirk thanks for review,
 If we don't  change here,  tests RexProgramTest#testSimplifyCondition and RexProgramTest#testSimplifyCondition2
will fail after we change the IS_NOT_NULL behavior in RexSimplify.
Yes, the piece of code is called really frequently and will have higher cost, so I wonder we should add a switch for these change.

Reply:
> we seem to be covered for this already by that the operators result is nullable even if all the inputs are non-nullable
> 
> ```java
>   @Test public void testUnsafeSimplify1() {
>     assertFalse(vIntNotNull(0).getType().isNullable());
>     assertTrue(plus(vIntNotNull(0), vIntNotNull(1)).getType().isNullable());
>     RexCall c= (RexCall)isNotNull(plus(vIntNotNull(0), vIntNotNull(1)));
>     assertFalse(c.isAlwaysFalse());
>     assertFalse(c.isAlwaysTrue());
>   }
> ```
the ' assertTrue(plus(vIntNotNull(0), vIntNotNull(1)).getType().isNullable());' case will fail,
that I miss something ？

Reply:
my bad...I probably reverted the changes from commandline and run the test from the ide...

Reply:
I still feel that this could bite back...
* I know it wont be 100% accurate: but I would feel that making the +/-* operators return value nullable regardless the arguments could help avoiding this recursive safetyness check.
* ...or annotate the nodes with their safety classification during construction; then we don't need to recurse to obtain this info